backend/src/Controllers/castVoteController.ts

```diff
@@ -0,0 +1,187 @@
+    const targetElectionId = req.params.id;
+    inputBallot.election_id = targetElectionId;
+    const targetElection = await ElectionsModel.getElectionByID(targetElectionId, req);
+    if (targetElection == null){
+        const errMsg = "Invalid Ballot: invalid election Id";
+        Logger.info(req, errMsg);
+        throw new BadRequest(errMsg);
+    }
+
+    if (targetElection.state!=='open'){
+        Logger.info(req, "Ballot Rejected. Election not open.", targetElection);
+        throw new BadRequest("Election is not open");
+    }
+
+    const voterId = getVoterID(req, targetElection.settings.voter_id_type, user);
+    Logger.debug(req, "voterID = " + voterId);
+    const roll = await getOrCreateElectionRoll(targetElection, voterId, req);
+    assertVoterMayVote(targetElection, roll, req);
+
+
+    //some ballot info should be server-authorative
+    inputBallot.date_submitted = Date.now();
+    if (inputBallot.history == null){
+        inputBallot.history = [];
+    }
+    inputBallot.ballot_id = randomUUID();
+    //TODO, ensure the user ID is added to the ballot...
+    //should server-authenticate the user id based on auth token
+    inputBallot.history.push({
+        action_type:"submit",
+        actor: voterId,
+        timestamp:inputBallot.date_submitted,
+    });
+
+    Logger.debug(req, "Submit Ballot:", inputBallot);
+    if (roll != null){
+        roll.ballot_id = String(inputBallot.ballot_id);
+        roll.submitted = true;
+        if (roll.history == null){
+            roll.history = [];
+        }
+        roll.history.push({
+            action_type:"submit",
+            actor: voterId,
+            timestamp:inputBallot.date_submitted,
+        });
+    }
+
+    const savedBallot = await persistBallotToStore(inputBallot, roll, req);
+    res.status("200").json({ ballot: savedBallot} );
+};
+
+
+async function persistBallotToStore(ballot:Ballot, roll:ElectionRoll|null, ctx:ILoggingContext):Promise<Ballot> {
+    Logger.debug(ctx, "persisteBallotToStore");
+    var savedBallot;
+    try {
+        if (roll == null){
+            savedBallot = await BallotModel.submitBallot(ballot, ctx, `User submits a ballot`);
+        } else {
+            savedBallot = await CastVoteStore.submitBallot(ballot, roll, ctx, 'User Submits a Ballot');
+        }
+        if (!savedBallot){
+            throw new InternalServerError("Failed to cast Ballot");
+        }
+    } catch (err:any){
+        Logger.error(ctx, 'Failed to cast Ballot:  '+err.message);
+        throw err;
+    }
+
+    return savedBallot;
+}
+
+
+function getVoterID(req:IRequest, voterIdType:string, user:any):string {
+    if (voterIdType === 'None') {
+        return "";
+    } else if (voterIdType === 'IP Address') {
+        Logger.debug(req, `ip=${String(req.ip)}`);
+        return String(req.ip)
+    } else if (voterIdType === 'Email') {
+        if (user == null){
+            throw new Unauthorized();
```

Comment:
Currently the front end is expecting a response with info about what is required to vote. Its probably better to respond with unauthorized, but we should probably add info about why its unauthorized like "Log in required" or "Voter ID reqired" or "Already voted"